File: crates/roost-linux/src/local_client.rs
//! In-process workspace adapter. Replaces the daemon-era
//! [`crate::client::RoostClient`] (gRPC) at M3b of the
//! daemon-removal refactor.
//!
//! `LocalClient` owns shared handles to a [`Workspace`] and a
//! [`PtySupervisor`] and exposes the small set of methods `app.rs`
//! invokes from its async-spawn closures. The shape mirrors the old
//! `RoostClient` so the call-sites in `app.rs` change minimally —
//! same method names, similar argument lists, results returning
//! `roost_ipc::messages` types (which have the same fields as the
//! retired proto types they replace).

use std::path::PathBuf;
use std::sync::Arc;

use anyhow::Result;
use roost_ipc::messages::{Project, Tab};
use tokio::sync::broadcast;

use crate::daemon::{PtyOutputEvent, PtySupervisor, Workspace};

/// In-process workspace + PTY supervisor handle.
#[derive(Clone)]
pub struct LocalClient {
    pub workspace: Arc<Workspace>,
    pub supervisor: Arc<PtySupervisor>,
    /// Socket path for `ROOST_SOCKET` env injection in spawned shells.
    pub socket_path: Arc<PathBuf>,
}

impl LocalClient {
    pub fn new(
        workspace: Arc<Workspace>,
        supervisor: Arc<PtySupervisor>,
        socket_path: PathBuf,
    ) -> Self {
        Self {
            workspace,
            supervisor,
            socket_path: Arc::new(socket_path),
        }
    }

    pub async fn list_projects(&self) -> Result<Vec<Project>> {
        Ok(self.workspace.snapshot())
    }

    pub async fn create_project(&self, name: &str, cwd: &str) -> Result<Project> {
        Ok(self.workspace.create_project(name, cwd)?)
    }

    pub async fn rename_project(&self, project_id: i64, name: &str) -> Result<()> {
        Ok(self.workspace.rename_project(project_id, name)?)
    }

    /// Delete a project and its tabs. Returns the cascaded tab ids
    /// so the caller can close the supervisor sessions.
    pub async fn delete_project(&self, project_id: i64) -> Result<Vec<i64>> {
        let cascaded = self.workspace.delete_project(project_id)?;
        for tab_id in &cascaded {
            self.supervisor.close(*tab_id);
        }
        Ok(cascaded)
    }

    pub async fn reorder_projects(&self, project_ids: Vec<i64>) -> Result<()> {
        Ok(self.workspace.reorder_projects(&project_ids)?)
    }

    pub async fn reorder_tabs(&self, project_id: i64, tab_ids: Vec<i64>) -> Result<()> {
        Ok(self.workspace.reorder_tabs(project_id, &tab_ids)?)
    }

    /// Open a tab and spawn the shell. Returns the tab snapshot
    /// plus a `broadcast::Receiver` subscribed BEFORE the supervisor's
    /// reader task started producing — `TabSession::attach_with_receiver`
    /// consumes it, no early-byte loss.
    pub async fn resize_tab(&self, tab_id: i64, cols: u32, rows: u32) -> Result<()> {
        // Same validation as `open_tab` — caller-supplied dims via
        // `roostctl tab resize` or via UI live-resize.
        let cols = pty_dim(cols, 80, "cols")?;
        let rows = pty_dim(rows, 24, "rows")?;
        self.supervisor
            .resize(tab_id, cols, rows)
            .await
            .map_err(|e| anyhow::anyhow!("pty resize failed: {e:?}"))
    }

    pub async fn open_tab(
        &self,
        project_id: i64,
        cwd: &str,
        title: &str,
        cols: u32,
        rows: u32,
    ) -> Result<(Tab, broadcast::Receiver<PtyOutputEvent>)> {
        // Resolve cwd: caller-supplied → project's cwd → $HOME →
        // "/". Matches the Mac side's `LocalClient.openTab`
        // fallback. Both the UI's open-new-tab path and the
        // `roostctl tab open` IPC call site land here, so the
        // fallback is centralized.
        let resolved_cwd: String = if !cwd.is_empty() {
            cwd.to_string()
        } else {
            let projects = self.workspace.snapshot();
            projects
                .into_iter()
                .find(|p| p.id == project_id)
                .map(|p| p.cwd)
                .filter(|c| !c.is_empty())
                .unwrap_or_else(|| std::env::var("HOME").unwrap_or_else(|_| "/".into()))
        };
        let tab = self.workspace.open_tab(project_id, &resolved_cwd, title)?;
        // Clamp + validate PTY dims. Zero → terminal default; values
        // exceeding u16 surface as a clear error rather than
        // silently truncating via `as u16` (CR-flagged: a CLI
        // caller passing `--cols 100000` would land with cols=34464
        // and a wildly-misshapen grid). Mirrors the Mac side's
        // `IPCHandlerImpl.ipcDim` validation.
        let cols = pty_dim(cols, 80, "cols")?;
        let rows = pty_dim(rows, 24, "rows")?;
        match self
            .supervisor
            .spawn(tab.id, &resolved_cwd, &[], cols, rows, &self.socket_path)
        {
            Ok(rx) => Ok((tab, rx)),
            Err(err) => {
                let _ = self.workspace.close_tab(tab.id);
                Err(anyhow::anyhow!("pty spawn failed: {err:?}"))
            }
        }
    }

    pub async fn close_tab(&self, tab_id: i64) -> Result<()> {
        self.supervisor.close(tab_id);
        Ok(self.workspace.close_tab(tab_id)?)
    }

    pub async fn set_tab_title(&self, tab_id: i64, title: &str) -> Result<()> {
        Ok(self.workspace.set_tab_title(tab_id, title)?)
    }

    /// Apply an OSC routing decision directly to the workspace.
    /// The legacy code path round-tripped this through the daemon
    /// via `ReportOsc`; in M3b the UI parses OSC in-process and
    /// updates state locally with no round-trip.
    pub fn apply_osc(&self, tab_id: i64, command: u32, payload: &str) {
        match command {
            0..=2 => {
                // Title set from the shell. OSC-from-shell path
                // never overrides a manual rename.
                let _ = self.workspace.set_tab_title_from_osc(tab_id, payload);
            }
            7 => {
                // OSC 7: cwd as `file://host/path` URI.
                if let Some(path) = parse_osc7_path(payload) {
                    let _ = self.workspace.set_tab_cwd(tab_id, &path);
                }
            }
            9 | 99 | 777 => {
                // Notification payload — surface to the UI via the
                // workspace's notification event. The actual
                // libnotify call happens in the UI layer once it
                // sees the WorkspaceEvent::NotificationFired event.
                let (title, body) = parse_notification_payload(command, payload);
                let _ = self.workspace.set_tab_has_notification(tab_id, true);
                let _ = self.workspace.fire_notification(tab_id, &title, &body);
            }
            _ => {
                tracing::debug!(tab_id, command, "ignored OSC");
            }
        }
    }
}

fn parse_osc7_path(payload: &str) -> Option<String> {
    // OSC 7 carries `file://host/abs/path`. The path portion starts
    // at the FIRST `/` after the host (or at index 0 if the host is
    // empty, e.g. `file:///tmp`). A malformed payload with no `/`
    // after the host returns None — the previous implementation's
    // `unwrap_or(0)` would have returned the host segment itself as
    // a "path," writing `host` into the workspace's cwd. CR-flagged.
    let after_scheme = payload.strip_prefix("file://")?;
    let path_start = after_scheme.find('/')?;
    Some(after_scheme[path_start..].to_string())
}

/// Validate + clamp a caller-supplied PTY dimension. Zero → the
/// supplied default; values exceeding `u16::MAX` return an error
/// instead of truncating via `as u16` (which would silently
/// produce e.g. cols=34464 for cols=100000). Mirrors the Rust
/// IPC handler's `u16::try_from` validation in `crates/roost-
/// linux/src/ipc.rs`.
fn pty_dim(value: u32, default: u16, field: &str) -> Result<u16> {
    if value == 0 {
        return Ok(default);
    }
    u16::try_from(value).map_err(|_| anyhow::anyhow!("{field} out of u16 range: {value}"))
}

fn parse_notification_payload(command: u32, payload: &str) -> (String, String) {
    match command {
        // OSC 777 ;notify;Title;Body — drop the leading `notify;`.
        777 => {
            let trimmed = payload.strip_prefix("notify;").unwrap_or(payload);
            let mut parts = trimmed.splitn(2, ';');
            let title = parts.next().unwrap_or("").to_string();
            let body = parts.next().unwrap_or("").to_string();
            (title, body)
        }
        // OSC 9 / 99 carry the title only.
        _ => (payload.to_string(), String::new()),
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn osc7_strips_host_prefix() {
        assert_eq!(
            parse_osc7_path("file://host/Users/me"),
            Some("/Users/me".into())
        );
    }

    #[test]
    fn osc7_handles_empty_host() {
        assert_eq!(parse_osc7_path("file:///tmp"), Some("/tmp".into()));
    }

    #[test]
    fn osc7_returns_none_for_host_without_path() {
        // `file://host` (no path after host) must not return "host"
        // as the path — that's the CR-flagged regression. Returns
        // None so the workspace cwd is left unchanged.
        assert_eq!(parse_osc7_path("file://host"), None);
    }

    #[test]
    fn osc777_splits_title_and_body() {
        assert_eq!(
            parse_notification_payload(777, "notify;Build;Passed"),
            ("Build".into(), "Passed".into())
        );
    }

    #[test]
    fn osc9_uses_payload_as_title() {
        assert_eq!(
            parse_notification_payload(9, "Hello"),
            ("Hello".into(), String::new())
        );
    }
}

File: mac/Sources/Roost/LocalClient.swift
// LocalClient.swift — daemon-removal refactor M4b.
//
// In-process adapter that the App will consume in M4b3 instead of
// `RoostClient` (gRPC). Wraps a shared `Workspace` + `PtySupervisor`
// + the IPC socket path so the same handles drive both the UI's
// state mutations and the IPC server's dispatch.
//
// Methods mirror `RoostClient`'s shape so the M4b3 rewire is a
// thin call-site rename per closure. Throws Swift-native errors
// (`Workspace.WorkspaceError`, `PtySupervisor.PtyError`) instead
// of gRPC `RPCError`s.

import Foundation

@MainActor
final class LocalClient {
    let workspace: Workspace
    let supervisor: PtySupervisor
    let socketPath: String

    init(workspace: Workspace, supervisor: PtySupervisor, socketPath: String) {
        self.workspace = workspace
        self.supervisor = supervisor
        self.socketPath = socketPath
    }

    // MARK: Projects

    func listProjects() -> [Workspace.Project] {
        workspace.snapshot()
    }

    @discardableResult
    func createProject(name: String, cwd: String) -> Workspace.Project {
        workspace.createProject(name: name, cwd: cwd)
    }

    func renameProject(_ projectID: Int64, name: String) throws {
        try workspace.renameProject(projectID, name: name)
    }

    /// Deletes the project and reaps every cascaded PTY. Returns
    /// the cascaded tab ids.
    @discardableResult
    func deleteProject(_ projectID: Int64) throws -> [Int64] {
        let cascaded = try workspace.deleteProject(projectID)
        for tabID in cascaded {
            supervisor.close(tabID: tabID)
        }
        return cascaded
    }

    func reorderProjects(_ projectIDs: [Int64]) throws {
        try workspace.reorderProjects(projectIDs)
    }

    func reorderTabs(projectID: Int64, tabIDs: [Int64]) throws {
        try workspace.reorderTabs(projectID: projectID, tabIDs: tabIDs)
    }

    // MARK: Tabs

    /// Open a tab and spawn the shell. The workspace records the
    /// tab first (which fires `tabOpened`); the supervisor then
    /// allocates the PTY. On supervisor failure the workspace tab
    /// is rolled back (fires `tabClosed`) and the error is
    /// rethrown.
    @discardableResult
    func openTab(
        projectID: Int64,
        cwd: String,
        argv: [String] = [],
        cols: UInt16 = 80,
        rows: UInt16 = 24,
        title: String = "",
        activate: Bool = true
    ) throws -> Workspace.Tab {
        let startCwd = Self.spawnCwd(
            requested: cwd,
            projectCwd: workspace.project(projectID)?.cwd ?? "",
            home: homeDirectory(),
            isDirectory: isDirectory
        )
        let tab = try workspace.openTab(
            projectID: projectID,
            cwd: startCwd,
            title: title,
            activate: activate
        )
        do {
            try supervisor.spawn(
                tabID: tab.id,
                cwd: startCwd,
                argv: argv,
                cols: cols,
                rows: rows,
                socketPath: socketPath
            )
        } catch {
            try? workspace.closeTab(tab.id)
            throw error
        }
        return tab
    }

    /// Where a shell asked to start in `requested` really starts: there
    /// if it is a directory, else in the project's cwd if that is one,
    /// else `home` — Rust's `application::usable_cwd`. `openTab` resolves
    /// it once and both records it in the row and hands it to the spawn,
    /// because a cwd that is not a directory would otherwise leave the
    /// child in Roost.app's own launch directory while the row names
    /// another (#541). An existing directory the shell cannot enter still
    /// counts, and fails the spawn.
    nonisolated static func spawnCwd(
        requested: String,
        projectCwd: String,
        home: String,
        isDirectory: (String) -> Bool
    ) -> String {
        [requested, projectCwd].first(where: isDirectory) ?? home
    }

    /// `tab.open`'s `project_id: 0`. A new project records the requested
    /// cwd only if it is a directory, and otherwise none, as Rust's does;
    /// its tabs then take the `$HOME` fallback when they spawn.
    @discardableResult
    func ensureDefaultProject(cwd: String, activate: Bool) -> Int64 {
        workspace.ensureDefaultProject(
            cwd: Self.spawnCwd(requested: cwd, projectCwd: "", home: "", isDirectory: isDirectory),
            activate: activate
        )
    }

    /// Where a tab opened from `tabID` starts — `tab.open`'s
    /// `cwd_from_tab`, mirroring Rust's `application::inherited_cwd`.
    func inheritedCwd(tabID: Int64) -> String? {
        guard let row = workspace.tab(tabID) else { return nil }
        return [supervisor.foregroundCwd(tabID: tabID), row.cwd]
            .compactMap { $0 }
            .first(where: isDirectory)
    }

    func closeTab(_ tabID: Int64) throws {
        // Row first, then the PTY, on both arms — the order
        // `roost_engine::application::close_tab` pins (#416). The main
        // actor already serializes this against the exit auto-close, so
        // on this side it is parity, not a fix.
        let removed = Result { try workspace.closeTab(tabID) }
        supervisor.close(tabID: tabID)
        try removed.get()
    }

    func setTabTitle(_ tabID: Int64, title: String) throws {
        try workspace.setTabTitle(tabID, title: title)
    }

    func setTabState(_ tabID: Int64, state: Workspace.TabState) throws {
        try workspace.setTabState(tabID, state: state)
    }

    func setTabHookActive(_ tabID: Int64, active: Bool) throws {
        try workspace.setTabHookActive(tabID, active: active)
    }

    @discardableResult
    func agentReport(_ report: AgentReport) throws -> (accepted: Bool, tab: Workspace.Tab) {
        try workspace.agentReport(report)
    }

    func clearTabNotification(_ tabID: Int64) throws {
        try workspace.setTabHasNotification(tabID, hasPending: false)
    }

    func focusTab(_ tabID: Int64) throws -> (previousProject: Int64, previousTab: Int64) {
        try workspace.focusTab(tabID)
    }

    /// A `false` return is suppression (plan §3.4 / §3.5), not failure —
    /// only a missing tab throws.
    @discardableResult
    func raiseAttention(
        _ tabID: Int64,
        title: String,
        body: String,
        source: Workspace.AttentionSource
    ) throws -> Bool {
        try workspace.raiseAttention(tabID, title: title, body: body, source: source)
    }

    // MARK: PTY I/O

    @discardableResult
    func writeTab(_ tabID: Int64, data: Data) throws -> Int {
        try supervisor.write(tabID: tabID, data: data)
    }

    func resizeTab(_ tabID: Int64, cols: UInt16, rows: UInt16) throws {
        try supervisor.resize(tabID: tabID, cols: cols, rows: rows)
    }

    // MARK: OSC routing

    /// Apply an OSC sequence directly to the workspace. Replaces
    /// the daemon-era `ReportOsc` round-trip. Called by the
    /// renderer's OSC callback in M4b3.
    func applyOSC(tabID: Int64, command: UInt32, payload: String) {
        switch command {
        case 0, 1, 2:
            // Shell-set title; respects a prior manual rename.
            try? workspace.setTabTitleFromOSC(tabID, title: payload)
        case 7:
            // OSC 7: cwd. The OSC scanner has already decoded
            // `file://host/path` → `/path` (see `OscEvent.asReport`'s
            // `.pwd` branch in `OscScanner.swift`), so `payload`
            // here is already a plain path. The earlier `if let
            // path = parseOSC7Path(payload)` re-parse expected
            // `file://...` and silently dropped the event because
            // the path no longer had that prefix. Pass through
            // verbatim — but defensively re-run `parseOSC7Path`
            // for the (theoretical) case where some external IPC
            // caller sends an unparsed URI through `tab.set_cwd`
            // or similar; the helper is idempotent on already-
            // parsed paths via the nil-on-no-scheme guard.
            let path = parseOSC7Path(payload) ?? payload
            if !path.isEmpty {
                try? workspace.setTabCwd(tabID, cwd: path)
            }
        case 9, 99, 777:
            // `rawOsc` is dropped while a live agent session is mid-turn:
            // the agent already reports its own attention through
            // `tab.agent_report`, and a wrapper shell echoing OSC 9 on
            // top of that double-notifies. The gate lives inside
            // `raiseAttention` alongside the mutation (plan §3.4).
            let (title, body) = parseNotificationPayload(command: command, payload: payload)
            _ = try? raiseAttention(tabID, title: title, body: body, source: .rawOsc)
        case 133:
            // OSC 133 prompt/command mark → the shell axis. Never gated:
            // the shell and agent axes are independent, and derivation
            // decides which one the tab shows.
            try? workspace.applyShellMark(tabID, body: payload)
        default:
            // Other OSC commands are ignored — the spec doesn't
            // route them to workspace state.
            break
        }
    }
}

private func isDirectory(_ path: String) -> Bool {
    var isDir: ObjCBool = false
    return FileManager.default.fileExists(atPath: path, isDirectory: &isDir) && isDir.boolValue
}

/// `$HOME` if it is an absolute path, else `/` — Rust's `home_dir` — so a
/// spawn cwd is never empty.
private func homeDirectory() -> String {
    let home = ProcessInfo.processInfo.environment["HOME"] ?? ""
    return home.hasPrefix("/") ? home : "/"
}

/// Strip the `file://` scheme + host segment from an OSC 7
/// payload. `file://host/path` → `/path`; `file:///abs` →
/// `/abs`; `file://hostonly` → nil (no path component, so we
/// don't overwrite the workspace cwd with a host token).
func parseOSC7Path(_ payload: String) -> String? {
    guard payload.hasPrefix("file://") else { return nil }
    let afterScheme = String(payload.dropFirst("file://".count))
    guard let slashIndex = afterScheme.firstIndex(of: "/") else { return nil }
    return String(afterScheme[slashIndex...])
}

/// Split an OSC 9 / 99 / 777 notification payload into
/// (title, body). OSC 777 strips a leading `notify;` then splits
/// title/body on the next `;`. OSC 9 / 99 use the entire payload
/// as the title.
func parseNotificationPayload(command: UInt32, payload: String) -> (String, String) {
    if command == 777 {
        let trimmed: String
        if payload.hasPrefix("notify;") {
            trimmed = String(payload.dropFirst("notify;".count))
        } else {
            trimmed = payload
        }
        if let sep = trimmed.firstIndex(of: ";") {
            return (
                String(trimmed[..<sep]),
                String(trimmed[trimmed.index(after: sep)...])
            )
        }
        return (trimmed, "")
    }
    return (payload, "")
}
